gilbreth_perception/src/voxelizer_node.cpp

```diff
@@ -0,0 +1,181 @@
+static const int voxel_grid_size = 32;
+
+ros::Publisher pub;
+
+//pcd to voxel functions
+struct Voxel {
+  unsigned int x;
+  unsigned int y;
+  unsigned int z;
+  Voxel() : x(0), y(0), z(0){};
+  Voxel(const unsigned int _x, const unsigned int _y, const unsigned int _z) : x(_x), y(_y), z(_z){};
+};
+
+template <typename PointT>
+const bool loadPointCloud(const std::string &file_path,
+                          typename pcl::PointCloud<PointT> &cloud_out) {
+  if (pcl::io::loadPCDFile<pcl::PointXYZ>(file_path.c_str(), cloud_out) == -1) {
+    PCL_ERROR("Failed to load PCD file \n");
+    return false;
+  }
+
+  return true;
+}
+
+// Get the linear index into a 1D array of voxels,
+// for a voxel at location (ix, iy, iz).
+const unsigned int getLinearIndex(const Voxel &voxel, const int grid_size) {
+  return voxel.x * (grid_size * grid_size) + voxel.z * grid_size + voxel.y;
+}
+
+const Voxel getGridIndex(const pcl::PointXYZ &point, const pcl::PointXYZ &translate, const uint voxel_grid_size, const float scale) {
+  // Needs to be signed to prevent overflow, because index can
+  // be slightly negative which then gets rounded to zero.
+  const int i = std::round(static_cast<float>(voxel_grid_size) * ((point.x - translate.x) / scale) - 0.5);
+  const int j = std::round(static_cast<float>(voxel_grid_size) * ((point.y - translate.y) / scale) - 0.5);
+  const int k = std::round(static_cast<float>(voxel_grid_size) * ((point.z - translate.z) / scale) - 0.5);
+  return Voxel(i, j, k);
+}
+
+// Format a float number to ensure it always has at least one decimal place
+// 0 --> 0.0
+// 1.1 --> 1.1
+// 1.10 --> 1.1
+const std::string formatFloat(const float value) {
+  std::stringstream ss;
+  ss << std::setprecision(6) << std::fixed << value;
+  std::string str;
+  ss.str().swap(str);
+  size_t last_zero_idx = str.find_last_not_of("0") + 1;
+  if (last_zero_idx == str.length()) {
+    // No trailing zeros
+    return str;
+  }
+  if (str[last_zero_idx - 1] == '.') {
+    // Last zero is after decimal point
+    last_zero_idx += 1;
+  }
+  str.resize(last_zero_idx);
+  return str;
+}
+
+void cloudCb(const sensor_msgs::PointCloud2ConstPtr &cloud_msg) {
+  pcl::PointCloud<pcl::PointXYZ>::Ptr scene(new pcl::PointCloud<pcl::PointXYZ>());
+  pcl::fromROSMsg(*cloud_msg, *scene);
+  //create and publish voxel
+  ros::Time start_time = ros::Time::now();
+  pcl::PointXYZ min_point;
+  pcl::PointXYZ max_point;
+  pcl::getMinMax3D(*scene, min_point, max_point);
+  // Calculate the scale factor so the longest side of the volume
+  // is split into the desired number of voxels
+  const float x_range = max_point.x - min_point.x;
+  const float y_range = max_point.y - min_point.y;
+  const float z_range = max_point.z - min_point.z;
+
+  const float max_cloud_extent = std::max(std::max(x_range, y_range), z_range);
+  const float voxel_size = max_cloud_extent / (static_cast<float>(voxel_grid_size) - 1.0);
+  std::cout << "voxel_size = " << voxel_size << std::endl;
```

Comment:
Use ROS_INFO or the appropriate ROS streaming macros.

Reply:
modified